### PR TITLE
Add module cv_server_provision for integration with Arista CloudVision Portal.

### PR DESCRIPTION
**Pull request by mharista**
##### SUMMARY
Adding first module for Ansible to integrate with Arista CloudVision Portal. The module allows a network team to export functionality from CloudVision Portal to be managed by another team that uses Ansible for server provisioning.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
modules/network/cloudvision/cv_server_provision

##### ANSIBLE VERSION
```
ansible 2.4.0 (cloudvision ac3906818f) last updated 2017/06/07 16:17:27 (GMT -400)
```

### CODE REVIEW

File: lib/ansible/modules/network/cloudvision/cv_server_provision.py [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: cv_server_provision
+version_added: "2.4"
+author: "EOS+ CS (ansible-dev@arista.com) (@mharista)"
+short_description:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudvision/cv_server_provision.py:31]: `short_description:` for modules should include the company & product name. It makes searching for modules easier

File: test/units/modules/network/cloudvision/test_cv_server_provision.py [hunk 2]
```diff
@@ -0,0 +1,891 @@
+#!/usr/bin/python
```
Comment (gundalow) [inline test/units/modules/network/cloudvision/test_cv_server_provision.py]: Please remove this line

File: lib/ansible/modules/network/cloudvision/cv_server_provision.py (hunk 1 above)
Comment (mharista) [reply lib/ansible/modules/network/cloudvision/cv_server_provision.py:31]: Added company and product name in commit d23dac6

**Comment by mharista** [comment]
Hi @gundalow,

I've tried a handful of things to resolve the E107 errors for the import statements but no luck. Is there something I am overlooking with how I have things formatted?

Thanks,
Matt

**Comment by AJamesPhillips** [comment]
@mharista one of the E107 errors might be referring to this space: https://github.com/ansible/ansible/pull/25450/commits/5e4498f2a84a1b0bc3eced8709e6161c8fd0686a#diff-32422873fd242eecf2efc56809d9350eR275

Other files I looked at don't have this.

** **edit** ** hmm but this one does: https://github.com/arista-eosplus/ansible/blob/5e4498f2a84a1b0bc3eced8709e6161c8fd0686a/lib/ansible/modules/files/archive.py#L116

**Comment by mharista** [comment]
Hi @AJamesPhillips,

Thanks for taking a look. I've tried running with that adjustment made (no change) and also seen similar examples of it working like the one you referenced.

**Comment by sivel** [comment]
The problem is in detecting the last line of any of `DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA`

This is due to using `\n` in the `sample`s from `RETURN`.

It is breaking line detection, and extending the last doc line to line 283.

Escaping those `\n` to make `\\n` or using literal YAML blocks such as:

```
newConfigBlock:
  description: The new config block for the user specified interface
  returned: when action = add or remove
  type: string
  sample: |
    interface Ethernet3
    description example
    no switchport
    !
``` 

Will resolve this problem.

There may some area for improvement in `validate-modules` but I won't be able to fix that immediately.  Please look to using one of the above "fixes" to pacify `validate-modules`.

**Comment by mharista** [comment]
Thanks @sivel!

**Comment by gundalow** [comment]
All the unit tests are failing

```
            ImportError while importing test module '/root/src/github.com/ansible/ansible/test/units/modules/network/cloudvision/test_cv_server_provision.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
test/units/modules/network/cloudvision/test_cv_server_provision.py:20: in <module>
    import ansible.modules.network.cloudvision.cv_server_provision as cv_server_provision
lib/ansible/modules/network/cloudvision/cv_server_provision.py:294: in <module>
    from cvprac.cvp_client import CvpClient
E   ImportError: No module named cvprac.cvp_client
```

**Comment by gundalow** [comment]
Assuming the library is installable via pip you can add it to `test/runner/requirements/units.txt`

**Comment by mharista** [comment]
Hi @gundalow,

Thanks for the help. My current failure is an import error when running the python 2.6 unit tests. Am I better off converting my unit tests to use unittest in ansible.compat.tests? I noticed other eos test modules using that.

**Comment by mattclay** [comment]
Have you considered mocking cvprac instead of importing it? That way we're only testing the module, rather than its dependencies.

File: test/units/modules/network/cloudvision/test_cv_server_provision.py [hunk 3]
```diff
@@ -0,0 +1,890 @@
...
+import time
+import ansible.modules.network.cloudvision.cv_server_provision as cv_server_provision
+import jinja2
+from jinja2 import meta
+from nose.tools import assert_equals, assert_raises, raises,\
+    assert_is_instance, assert_is_not_none, assert_is_none,\
+    assert_true, assert_false, assert_not_in
+
+from ansible.module_utils.basic import AnsibleModule
+from cvprac.cvp_client import CvpClient
```
Comment (gundalow) [inline test/units/modules/network/cloudvision/test_cv_server_provision.py]: To allow people to run the unit tests locally that don't have the dependencies installed (i.e. outside of Shippable) could you please update the imports to match https://github.com/ansible/ansible/pull/25762

File: test/runner/requirements/units.txt [hunk 4]
```diff
@@ -21,3 +21,6 @@ ipaddress
 f5-sdk ; python_version >= '2.7'
 f5-icontrol-rest ; python_version >= '2.7'
 deepdiff
+
+# requirement for CVP specific modules
```
Comment (gundalow) [inline test/runner/requirements/units.txt]: `# Needed for modules/network/cloudvision/`

**Comment by mharista** [comment]
@mattclay I ended up mocking all references to CvpClient in addition to reformatting the tests.

**Comment by gundalow** [comment]
Looks like the `logging` module that's included in `2017-06-17 13:19:24 Collecting logging (from cvprac->-r test/runner/requirements/units.txt (line 26))` is conflicting with the normal version on Python 2.5 & 3.6

```
2017-06-17 13:19:25       File "<string>", line 1, in <module>
2017-06-17 13:19:25       File "/tmp/pip-build-zq324jpm/logging/setup.py", line 13, in <module>
2017-06-17 13:19:25         packages = ["logging"],
2017-06-17 13:19:25       File "/usr/lib/python3.5/distutils/core.py", line 148, in setup
2017-06-17 13:19:25         dist.run_commands()
2017-06-17 13:19:25       File "/usr/lib/python3.5/distutils/dist.py", line 955, in run_commands
2017-06-17 13:19:25         self.run_command(cmd)
2017-06-17 13:19:25       File "/usr/lib/python3.5/distutils/dist.py", line 974, in run_command
2017-06-17 13:19:25         cmd_obj.run()
2017-06-17 13:19:25       File "/root/src/github.com/ansible/ansible/test/runner/.tox/py35/lib/python3.5/site-packages/setuptools/command/egg_info.py", line 280, in run
2017-06-17 13:19:25         self.find_sources()
2017-06-17 13:19:25       File "/root/src/github.com/ansible/ansible/test/runner/.tox/py35/lib/python3.5/site-packages/setuptools/command/egg_info.py", line 295, in find_sources
2017-06-17 13:19:25         mm.run()
2017-06-17 13:19:25       File "/root/src/github.com/ansible/ansible/test/runner/.tox/py35/lib/python3.5/site-packages/setuptools/command/egg_info.py", line 526, in run
2017-06-17 13:19:25         self.add_defaults()
2017-06-17 13:19:25       File "/root/src/github.com/ansible/ansible/test/runner/.tox/py35/lib/python3.5/site-packages/setuptools/command/egg_info.py", line 562, in add_defaults
2017-06-17 13:19:25         sdist.add_defaults(self)
2017-06-17 13:19:25       File "/root/src/github.com/ansible/ansible/test/runner/.tox/py35/lib/python3.5/site-packages/setuptools/command/py36compat.py", line 34, in add_defaults
2017-06-17 13:19:25         self._add_defaults_python()
2017-06-17 13:19:25       File "/root/src/github.com/ansible/ansible/test/runner/.tox/py35/lib/python3.5/site-packages/setuptools/command/sdist.py", line 133, in _add_defaults_python
2017-06-17 13:19:25         build_py = self.get_finalized_command('build_py')
2017-06-17 13:19:25       File "/usr/lib/python3.5/distutils/cmd.py", line 298, in get_finalized_command
2017-06-17 13:19:25         cmd_obj = self.distribution.get_command_obj(command, create)
2017-06-17 13:19:25       File "/usr/lib/python3.5/distutils/dist.py", line 846, in get_command_obj
2017-06-17 13:19:25         klass = self.get_command_class(command)
2017-06-17 13:19:25       File "/root/src/github.com/ansible/ansible/test/runner/.tox/py35/lib/python3.5/site-packages/setuptools/dist.py", line 493, in get_command_class
2017-06-17 13:19:25         self.cmdclass[command] = cmdclass = ep.load()
2017-06-17 13:19:25       File "/root/src/github.com/ansible/ansible/test/runner/.tox/py35/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2302, in load
2017-06-17 13:19:25         return self.resolve()
2017-06-17 13:19:25       File "/root/src/github.com/ansible/ansible/test/runner/.tox/py35/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2308, in resolve
2017-06-17 13:19:25         module = __import__(self.module_name, fromlist=['__name__'], level=0)
2017-06-17 13:19:25       File "/root/src/github.com/ansible/ansible/test/runner/.tox/py35/lib/python3.5/site-packages/setuptools/command/build_py.py", line 15, in <module>
2017-06-17 13:19:25         from setuptools.lib2to3_ex import Mixin2to3
2017-06-17 13:19:25       File "/root/src/github.com/ansible/ansible/test/runner/.tox/py35/lib/python3.5/site-packages/setuptools/lib2to3_ex.py", line 12, in <module>
2017-06-17 13:19:25         from lib2to3.refactor import RefactoringTool, get_fixers_from_package
2017-06-17 13:19:25       File "/usr/lib/python3.5/lib2to3/refactor.py", line 19, in <module>
2017-06-17 13:19:25         import logging
2017-06-17 13:19:25       File "/tmp/pip-build-zq324jpm/logging/logging/__init__.py", line 618
2017-06-17 13:19:25         raise NotImplementedError, 'emit must be implemented '\
2017-06-17 13:19:25                                  ^
2017-06-17 13:19:25     SyntaxError: invalid syntax
```
https://stackoverflow.com/questions/32386469/logging-module-not-working-with-python3

**Comment by mharista** [comment]
I'm trying to mock cvprac entirely from the unit tests but am still having issues. I'm also seeing that python 3.5 mock doesn't have the assert_called_once method but it is added back in for python 3.6. Is it mandatory to support all four versions of python (2.6, 2.7, 3.5, 3.6)?

**Comment by gundalow** [comment]
@mharista Code generally must work with py2 + py3.

It's fine if a library doesn't support py 2.6 to skip that

**Comment by gundalow** [comment]
Maybe skip the test on py3.5.

As long as the module works on py3.5 & py 3.6 that should be good enough

**Comment by mharista** [comment]
@gundalow Thanks for the help.

I decided to convert all unit tests away from using the assert_called_once method, to support python 3.5, so that all the checks could pass for all versions.

**Comment by mharista** [comment]
Hi @gundalow,

Is there anything else I need to address here or are all the shippable tests passing sufficient?

Thanks,
Matt

**Comment by gundalow** [comment]
closing and reopening to trigger CI

**Comment by mharista** [comment]
Hi @gundalow,

After the close/open of the PR ansible-test is now reporting the import errors for jinja2. Is this expected? Is there another way Ansible handles interaction with Jinja2 templates within a module?

**Comment by gundalow** [comment]
@mharista We've recently improved the CI detecting broken imports, you just need to do

```
try: 
  import jinja2
  from jinja2 import meta
  HAS_JINHA2 = TRUE
except ImportError:
  HAS_JINJA2 = False
```

Then in your `main`

```
if not HAS_JINJA2:
  module.fail_json(msg='Jinja2 is needed on the remote host for this module to function')
```


For an example in another module see
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/commands/expect.py#L94-L98

https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/commands/expect.py#L132-L133

**Comment by mharista** [comment]
@gundalow I reformatted the imports for Jinja2 and cvprac but am now hitting a sanity error in the boilerplate script. Is this something I can skip as referenced here - http://docs.ansible.com/ansible/latest/dev_guide/testing_sanity.html

**Comment by gundalow** [comment]
Have a look at https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/commands/command.py#L9-L10

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Merged, and will be released in Ansible 2.4.0

**Pull request merged**